File: src/llm/assistant.js
import axios from "axios";
import config from "../config.js";
import { systemPrompt } from "./prompt.js";
import { mockReply } from "./mock.js";

// ─────────────────────────────────────────────────────────────
// Admission Knowledge Assistant.
// reply(history, context) -> { reply, leadUpdates, endCall }
//
// history: [{ role: "user"|"assistant", content: string }, ...]
// Providers: anthropic | openai-compatible | mock (offline, KB-driven).
// ─────────────────────────────────────────────────────────────

function parseModelJson(text) {
  if (!text) return null;
  // Strip code fences / surrounding prose and grab the first {...} block.
  const match = text.match(/\{[\s\S]*\}/);
  if (!match) return { reply: text.trim(), lead_updates: {}, end_call: false };
  try {
    return JSON.parse(match[0]);
  } catch {
    return { reply: text.trim(), lead_updates: {}, end_call: false };
  }
}

async function anthropicReply(sys, history) {
  const { data } = await axios.post(
    "https://api.anthropic.com/v1/messages",
    {
      model: config.llm.model,
      max_tokens: 160,
      system: sys,
      messages: history.map((m) => ({ role: m.role, content: m.content })),
    },
    {
      headers: {
        "x-api-key": config.llm.apiKey,
        "anthropic-version": "2023-06-01",
        "Content-Type": "application/json",
      },
      timeout: 15000,
    }
  );
  return data?.content?.[0]?.text || "";
}

async function openaiReply(sys, history) {
  const base = config.llm.baseUrl || "https://api.openai.com/v1";
  const { data } = await axios.post(
    `${base.replace(/\/$/, "")}/chat/completions`,
    {
      model: config.llm.model,
      max_tokens: 160,
      temperature: 0.6,
      messages: [{ role: "system", content: sys }, ...history],
    },
    {
      headers: {
        Authorization: `Bearer ${config.llm.apiKey}`,
        "Content-Type": "application/json",
      },
      timeout: 15000,
    }
  );
  return data?.choices?.[0]?.message?.content || "";
}

export async function reply(history, ctx = {}) {
  const provider = config.llm.provider;

  // Offline / no-key path: rule-based KB assistant (keeps the demo working).
  if (provider === "mock" || !config.llm.apiKey) {
    return mockReply(history, ctx);
  }

  const sys = systemPrompt(ctx);
  try {
    let raw = "";
    if (provider === "anthropic") raw = await anthropicReply(sys, history);
    else raw = await openaiReply(sys, history);

    const parsed = parseModelJson(raw) || {};
    return {
      reply: parsed.reply || "માફ કરશો, ફરી એકવાર કહેશો?",
      leadUpdates: parsed.lead_updates || {},
      endCall: Boolean(parsed.end_call),
    };
  } catch (err) {
    console.error("[llm] provider error, falling back to mock:", err.message);
    return mockReply(history, ctx);
  }
}

File: app/llm/assistant.py
"""Admission knowledge assistant — LLM dispatch (ported from src/llm/assistant.js).

reply(history, ctx) -> {"reply", "leadUpdates", "endCall"}
Providers: anthropic | openai-compatible | mock (offline, KB-driven).
"""
import json
import re

import httpx

from ..config import config
from .prompt import system_prompt
from .mock import mock_reply


def _parse_model_json(text):
    if not text:
        return None
    m = re.search(r"\{[\s\S]*\}", text)
    if not m:
        return {"reply": text.strip(), "lead_updates": {}, "end_call": False}
    try:
        return json.loads(m.group(0))
    except Exception:
        return {"reply": text.strip(), "lead_updates": {}, "end_call": False}


async def _anthropic_reply(sys, history):
    async with httpx.AsyncClient(timeout=10.0) as client:
        r = await client.post(
            "https://api.anthropic.com/v1/messages",
            json={
                "model": config.llm.model,
                "max_tokens": 160,
                "system": sys,
                "messages": [{"role": m["role"], "content": m["content"]} for m in history],
            },
            headers={
                "x-api-key": config.llm.api_key,
                "anthropic-version": "2023-06-01",
                "Content-Type": "application/json",
            },
        )
        r.raise_for_status()
        data = r.json()
        content = data.get("content") or [{}]
        return content[0].get("text", "")


async def _openai_reply(sys, history):
    base = (config.llm.base_url or "https://api.openai.com/v1").rstrip("/")
    async with httpx.AsyncClient(timeout=10.0) as client:
        r = await client.post(
            f"{base}/chat/completions",
            json={
                "model": config.llm.model,
                "max_tokens": 160,
                "temperature": 0.6,
                "messages": [{"role": "system", "content": sys}, *history],
            },
            headers={"Authorization": f"Bearer {config.llm.api_key}", "Content-Type": "application/json"},
        )
        r.raise_for_status()
        data = r.json()
        return (data.get("choices") or [{}])[0].get("message", {}).get("content", "")


async def reply(history, ctx=None):
    ctx = ctx or {}
    provider = config.llm.provider

    if provider == "mock" or not config.llm.api_key:
        return mock_reply(history, ctx)

    sys = system_prompt(ctx.get("direction", "inbound"), ctx.get("studentName", ""))
    try:
        if provider == "anthropic":
            raw = await _anthropic_reply(sys, history)
        else:
            raw = await _openai_reply(sys, history)
        parsed = _parse_model_json(raw) or {}
        return {
            "reply": parsed.get("reply") or "માફ કરશો, ફરી એકવાર કહેશો?",
            "leadUpdates": parsed.get("lead_updates") or {},
            "endCall": bool(parsed.get("end_call")),
        }
    except Exception as e:
        print("[llm] provider error, falling back to mock:", str(e))
        return mock_reply(history, ctx)
